### PR TITLE
JIT: Disqualify methods with modifiable `this` from OSR.

### LINKED ISSUE #72845: There are circumstances where ListEquality is broken on net7.0
There are circumstances where ListEquality is broken on net7.0

Repro:
1. install a dotnet 7.0
2. unzip this file 
[repro.zip](https://github.com/dotnet/fsharp/files/9185052/repro.zip)
3. cd to where you unzipped it:
4. Type dotnet run
5. Observe the output:
C:\temp\repro>dotnet run
Version: .NET 7.0.0-preview.7.22369.4
**failure: vwnwer-e**

Then either uninstall net7.0 or rename the directory containing the shared runtime: 
C:\Program Files\dotnet\shared\Microsoft.NETCore.App\7.0.0-preview.7.22369.4 to something like:
C:\Program Files\dotnet\shared\Microsoft.NETCore.App\XXXXXX7.0.0-preview.7.22369.4

6. And type dotnet run
7. observe the output:
C:\temp\repro>dotnet run
Version: .NET 6.0.7
**success: vwnwer-e**

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Disqualify methods with modifiable `this` from OSR.

Closes #72845.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@jakobbotsch PTAL
cc @dotnet/jit-contrib 

Codegen diff with the fix -- OSR method fetches the `this` from a different Tier0 slot.
```diff
@@ -190,30 +190,30 @@ G_M36567_IG01:              ;; offset=0000H
        4889BC24E8000000     mov      qword ptr [rsp+E8H], rdi
        4889B424E0000000     mov      qword ptr [rsp+E0H], rsi
        48899C24D8000000     mov      qword ptr [rsp+D8H], rbx
-       488B8C2400010000     mov      rcx, gword ptr [rsp+100H]
+       488B4C2478           mov      rcx, gword ptr [rsp+78H]
```

File: src/coreclr/jit/compiler.cpp [hunk 1]
```diff
@@ -5360,6 +5360,23 @@ void Compiler::generatePatchpointInfo()
...
+        // the potential modification by Tier0 vis-a-vis the timing of any OSR transition.
+        //
+        // Note when the orginal `this` must be kept alive (for say generics context
+        // reporting) its location is conveyed by a separate field in the patchpoint info.
+        // The OSR method will not need to refer to the original Tier0 `this` in codegen.
+        //
+        if ((lclNum == 0) && !info.compIsStatic && !lvaIsOriginalThisReadOnly())
+        {
+            varNum = lvaArg0Var;
+        }
```
Comment (jakobbotsch) [inline src/coreclr/jit/compiler.cpp:5378]: Is there a risk that this ends up changing the generic context if we are fetching that through the this pointer? Will the OSR version suddenly see a new generic context?

Comment (jakobbotsch) [reply src/coreclr/jit/compiler.cpp:5378]: For example:
```csharp
class CBase<T>
{
  void Foo()
  {
    this = new CDerived<string, T>();
    // OSR triggers here
    Console.WriteLine(typeof(T).FullName);
  }
}

class CDerived<T1, T2> : CBase<T2>
{
}
```

May need some finessing...

Comment (AndyAyersMS) [reply src/coreclr/jit/compiler.cpp:5378]: Yes, seems like that could happen, `getRuntimeContextTree` explicitly refers to the unmodified `this`.

Might require IL to create a suitable example.

Comment (jakobbotsch) [reply src/coreclr/jit/compiler.cpp:5378]: `[MethodImpl(MethodImplOptions.Synchronized)]` may have the same problem if that is an easier test to write.

Comment (AndyAyersMS) [reply src/coreclr/jit/compiler.cpp:5378]: I haven't come up with anything yet. C# doesn't allow assignment to this (for ref class instance methods anyways).

I suppose I could look at the F# example more closely.

**Comment by AndyAyersMS** [comment]
Also looks like I'll have to exclude this new test from mono aot.

**Comment by AndyAyersMS** [comment]
Decided bailing out on OSR was the safer and less messy option.

File: src/coreclr/jit/compiler.hpp [hunk 2]
```diff
@@ -4703,6 +4703,10 @@ inline bool Compiler::compCanHavePatchpoints(const char** reason)
     {
         whyNot = "OSR can't handle reverse pinvoke";
     }
+    else if (!info.compIsStatic && !lvaIsOriginalThisReadOnly())
+    {
+        whyNot = "OSR can't handle modifiable this";
+    }
```
Comment (jakobbotsch) [inline src/coreclr/jit/compiler.hpp:4709]: What happens to these methods with tiering enabled? Do they end up in tier 0 or tier 1?
It seems risky if tail-recursive F# functions that are turned into loops by the compiler end up stuck in tier 0.

Comment (AndyAyersMS) [reply src/coreclr/jit/compiler.hpp:4709]: `compCompileHelper` has the logic for this -- methods with explicit `tail` prefixes get optimized unless we're in PGO mode.

https://github.com/dotnet/runtime/blob/56a1cf4c892c6dec1d462e08e177f470e2f6b7b3/src/coreclr/jit/compiler.cpp#L6644-L6715

Comment (jakobbotsch) [reply src/coreclr/jit/compiler.hpp:4709]: They are not going to have any tail prefixes. I think #72845 is a case of a tail recursive function on lists that the F# compiler is turning into a loop in IL by reassigning `this`. These kind of methods may be quite common in F# code so getting them stuck in tier 0 could be problematic.

Comment (AndyAyersMS) [reply src/coreclr/jit/compiler.hpp:4709]: If methods have explicit loops and modify `this` then they'll be disqualified from OSR by the logic being added to `compCanHavePatchpoints`, so will end up getting optimized.

For example, `Equals` in this test case now gets optimized immediately.

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/compiler.hpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/compiler.hpp:4709]: Thanks, then it looks good.

**Comment by AndyAyersMS** [comment]
Mono LLFVAOT timed out, will ignore.

**Pull request merged**